File: leetcode/binary_tree/average_of_levels.rb
def average_of_levels(root)
  return [] if root.nil?

  curr_q = [root]
  levels = []

  while !curr_q.empty?
    next_q = []
    sum = 0
    l = 0

    while !curr_q.empty?
      node = curr_q.shift
      sum += node.val
      l += 1

      next_q << node.left if node.left
      next_q << node.right if node.right
    end

    levels << sum / l.to_f
    curr_q = next_q
  end

  levels
end

File: leetcode/binary_tree/averageOfLevels.js
var averageOfLevels = function(root) {
  const results = [];
  let parents = [root];

  while (parents.length > 0) {
    const len = parents.length;
    const sum = parents.map(p => p.val).reduce((a, b) => a + b);
    let children = [];

    results.push(sum / len);

    while (parents.length > 0) {
      const parent = parents.shift();

      if (parent.left) children.push(parent.left);
      if (parent.right) children.push(parent.right);
    }

    parents = children;
  }

  return results;
};
